### ISSUE TITLE
Problem with mfc1 emulation

### ISSUE THREAD

**Issue by douggilmore** [prplfoundation/qemu#32]
We've noticed a problem with QEMU that seems to be exposed by some
MIPS32R6 compilers (unfortunately not ToT) when running
the Golden Oldie version of Linpack (C) on QEMU.

To reproduce (we are working on the logistics of getting a tool chain
out to reproduce the problem):

$ wget http://www.netlib.org/benchmark/linpackc
--2015-01-13 14:33:57--  http://www.netlib.org/benchmark/linpackc
Resolving www.netlib.org (www.netlib.org)... 160.36.131.221
Connecting to www.netlib.org (www.netlib.org)|160.36.131.221|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 21049 (21K) [text/plain]
Saving to: `linpackc.1'

100%[======================================>] 21,049      --.-K/s   in 0.07s   

2015-01-13 14:33:57 (291 KB/s) - `linpackc.1' saved [21049/21049]

$ cp linpackc linpack.c
$ cat od
--- linpackc	2009-08-27 09:02:55.000000000 -0700
+++ linpack.c	2015-01-12 11:04:19.917866000 -0800
@@ -896,7 +896,9 @@
 struct rusage ru;
 REAL t ;
 
+#if 0
 void getrusage();
+#endif
  
 getrusage(RUSAGE_SELF,&ru) ;
  
$ patch linpack.c od
patching file linpack.c
$ mips-img-linux-gnu-gcc -EL -DDP -DROLL -DMIPS -g -O2 linpack.c -mmsa -lm -static -save-temps -funroll-all-loops -o lp
$ /home/dgilmore/qemu/bg27149d8-db/local/bin/qemu-mipsel -cpu MIPS32R6-generic -d all -D oq ./lp 20 1
Rolled Double Precision Linpack

Rolled Double Precision Linpack

     norm. resid      resid           machep         x[0]-1        x[n-1]-1
       1.9        8.46778499e-14  2.22044605e-16 -1.11799459e-13 -9.60342916e-14
    times are reported for matrices of order   100
      dgefa      dgesl      total       kflops     unit      ratio
 times for array with leading dimension of  201
       2.08       0.06       2.14        321       6.22      38.15
qemu: uncaught target signal 11 (Segmentation fault) - core dumped
Segmentation fault (core dumped)
$ 

Note that the end of the log file oq contains:

...
IN: dgefa
0x00404118:  mfc1	v0,$f26
0x0040411c:  b	0x403e2c
0x00404120:  ldc1	$f0,0(v0)
...

Running with -singlestep will finesse the problem.

**Comment by douggilmore** [comment]
For a copy of the (gziped) executable see the page:

https://dmz-portal.mips.com/wiki/MFC1_issue_in_Qemu_that_is_exposed_by_R6

And click on:

File:Lp.gz

**Comment by douggilmore** [comment]
md5sum information:

352d3172252f9d0137d9fd9090df4608  Lp.gz

**Comment by lalrae** [comment]
I reproduced the problem in my small userland program. Problematic combination is mfc1 followed by any load/store instruction which uses returned value as an address. Example:
0x00400ab8:  mfc1  t2,$f2
0x00400abc:  lw    s4,0(t2)

Content of $f2:
f2 = 0x0000abcd76fff000

MFC1 moves word from FPR $f2 into GPR $t2. After executing this instruction the content of GPR $t2 is:
t2 = 0x76fff000

This seems to look good. However, during TCG optimizations in this translation block "lw" instruction actually uses the whole content of FPR $f2 rather than truncated 32-bit value from GPR $t2. That means if upper 32 bits of FPR $f2 aren't zero (in above example it is 0xabcd) then different address will be referenced, hence the guest program crashes.

I haven't found anything suspicious in TCG code generation in MIPS front-end for these instructions. I'm suspecting something is going wrong during TCG optimization step. What I can see is that this problem does NOT occur when:
1) Disabling TCG optimizations.
2) Running in singlestep mode. "mfc1" and "lw" are placed in separate translation blocks, thus the faulty optimization doesn't happen.
3) Running the same instruction combination in system emulation. Probably because back-end of load/store operations is different than in userland emulation (we need to handle TLB etc.).

I also tried to combine "mfc1" with other instructions than load/store, but I haven't noticed any malfunction. So for now it looks like only "mfc1" + load/store in userland emulation has this problem.

This requires further investigation, for now I can provide a workaround where translation block is terminated on mfc1 to avoid faulty TCG optimization. I will send the patch shortly.
